src/platform/nrfconnect/wifi/WiFiManager.cpp

```diff
@@ -0,0 +1,447 @@
+    Inet::UDPEndPointImplSockets::SetJoinMulticastGroupHandler([](Inet::InterfaceId interfaceId, const Inet::IPAddress & address) {
+        const in6_addr addr = InetUtils::ToZephyrAddr(address);
+        net_if * iface      = InetUtils::GetInterface(interfaceId);
+        VerifyOrReturnError(iface != nullptr, INET_ERROR_UNKNOWN_INTERFACE);
+
+        net_if_mcast_addr * maddr = net_if_ipv6_maddr_add(iface, &addr);
+
+        if (maddr && !net_if_ipv6_maddr_is_joined(maddr) && !net_ipv6_is_addr_mcast_link_all_nodes(&addr))
+        {
+            net_if_ipv6_maddr_join(maddr);
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    Inet::UDPEndPointImplSockets::SetLeaveMulticastGroupHandler([](Inet::InterfaceId interfaceId, const Inet::IPAddress & address) {
+        const in6_addr addr = InetUtils::ToZephyrAddr(address);
+        net_if * iface      = InetUtils::GetInterface(interfaceId);
+        VerifyOrReturnError(iface != nullptr, INET_ERROR_UNKNOWN_INTERFACE);
+
+        if (!net_ipv6_is_addr_mcast_link_all_nodes(&addr) && !net_if_ipv6_maddr_rm(iface, &addr))
+        {
+            return CHIP_ERROR_INVALID_ADDRESS;
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    ChipLogDetail(DeviceLayer, "wpa_supplicant has been initialized");
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR WiFiManager::AddNetwork(const ByteSpan & ssid, const ByteSpan & credentials)
+{
+    ChipLogDetail(DeviceLayer, "Adding WiFi network");
+    mpWpaNetwork = wpa_supplicant_add_network(wpa_s);
+    if (mpWpaNetwork)
+    {
+        static constexpr size_t kMaxSsidLen{ 32 };
+        mpWpaNetwork->ssid = (u8 *) k_malloc(kMaxSsidLen);
+
+        if (mpWpaNetwork->ssid)
+        {
+            memcpy(mpWpaNetwork->ssid, ssid.data(), ssid.size());
+            mpWpaNetwork->ssid_len    = ssid.size();
+            mpWpaNetwork->key_mgmt    = WPA_KEY_MGMT_NONE;
+            mpWpaNetwork->disabled    = 1;
+            wpa_s->conf->filter_ssids = 1;
+
+            return AddPsk(credentials);
+        }
+    }
+
+    return CHIP_ERROR_INTERNAL;
+}
+
+CHIP_ERROR WiFiManager::Scan(const ByteSpan & ssid, ScanCallback callback)
+{
+    const StationStatus stationStatus = GetStationStatus();
+    VerifyOrReturnError(stationStatus != StationStatus::DISABLED && stationStatus != StationStatus::SCANNING &&
+                            stationStatus != StationStatus::CONNECTING,
+                        CHIP_ERROR_INCORRECT_STATE);
+
+    net_if * const iface = InetUtils::GetInterface();
+    VerifyOrReturnError(iface != nullptr, CHIP_ERROR_INTERNAL);
+
+    const device * dev = net_if_get_device(iface);
+    VerifyOrReturnError(dev != nullptr, CHIP_ERROR_INTERNAL);
+
+    const net_wifi_mgmt_offload * ops = static_cast<const net_wifi_mgmt_offload *>(dev->api);
+    VerifyOrReturnError(ops != nullptr, CHIP_ERROR_INTERNAL);
+
+    mScanCallback = callback;
+
+    // TODO: Use saner API once such exists.
```

Comment:
Hmm. Marcin has already switched that to a new management API. Do we use older one because it requires NCS 2.2?

Reply:
Oh, I've realized right now that I should align upstream to the NCS 2.1.2 instead of NCS 2.1.1. So, I will add these commits.